src/phoenix/experimental/callbacks/llama_index_callback_handler.py

```diff
@@ -0,0 +1,163 @@
+        **kwargs: Any,
+    ) -> CBEventID:
+        event_id = event_id or str(uuid4())
+        self._event_id_to_event_data[event_id]["name"] = event_type.value
+        self._event_id_to_event_data[event_id]["event_type"] = event_type
+        self._event_id_to_event_data[event_id]["start_event"] = CBEvent(
+            event_type=event_type,
+            payload=payload,
+            id_=event_id,
+        )
+        return event_id
+
+    def on_event_end(
+        self,
+        event_type: CBEventType,
+        payload: Optional[Dict[str, Any]] = None,
+        event_id: CBEventID = "",
+        **kwargs: Any,
+    ) -> None:
+        self._event_id_to_event_data[event_id].setdefault("name", event_type.value)
+        self._event_id_to_event_data[event_id].setdefault("event_type", event_type)
+        self._event_id_to_event_data[event_id]["end_event"] = CBEvent(
+            event_type=event_type,
+            payload=payload,
+            id_=event_id,
+        )
+
+    def start_trace(self, trace_id: Optional[str] = None) -> None:
+        self._event_id_to_event_data = defaultdict(lambda: CBEventData())
+
+    def end_trace(
+        self,
+        trace_id: Optional[str] = None,
+        trace_map: Optional[Dict[CBEventID, List[CBEventID]]] = None,
+    ) -> None:
+        if not trace_map:
+            return  # TODO: investigate when empty or None trace_map is passed
+        try:
+            _add_to_tracer(
+                event_id_to_event_data=self._event_id_to_event_data,
+                trace_map=trace_map,
+                tracer=self._tracer,
+            )
+        except Exception:
+            logger.exception("OpenInferenceCallbackHandler trace processing failed")
+        self._event_id_to_event_data = defaultdict(lambda: CBEventData())
+
+
+def _add_to_tracer(
+    event_id_to_event_data: Dict[CBEventID, CBEventData],
+    trace_map: Dict[CBEventID, List[CBEventID]],
+    tracer: Tracer,
+) -> None:
+    """Adds event data to the tracer, where it is converted to a span and stored in a buffer.
+
+    Args:
+        event_id_to_event_data (Dict[CBEventID, CBEventData]): A map of event IDs to event data.
+
+        trace_map (Dict[CBEventID, List[CBEventID]]): A map of parent event IDs to child event IDs.
+        The root event IDs are stored under the key "root".
+
+        tracer (Tracer): The tracer that stores spans.
+    """
+
+    parent_child_id_stack: List[Tuple[Optional[SpanID], CBEventID]] = [
+        (None, root_event_id) for root_event_id in trace_map["root"]
+    ]
+    while parent_child_id_stack:
+        parent_span_id, event_id = parent_child_id_stack.pop()
+        event_data = event_id_to_event_data[event_id]
+        start_event = event_data["start_event"]
+        end_event = event_data["end_event"]
+        name = event_data["name"]
+        event_type = event_data["event_type"]
+        span_kind = _get_span_kind(event_type)
+        span = tracer.create_span(
```

Comment:
trace_id is missing here and in the function input

Reply:
I think this is what we want in this case. The `trace_id` parameter in the LlamaIndex callback system is a misnomer. It's a name (e.g., "query" or "index_construction") and not a unique identifier.

Reply:
I'm letting the tracer create the trace ID

Reply:
Whoops, you're right. Addressed in [52c589f](https://github.com/Arize-ai/phoenix/pull/1058/commits/52c589fa743a2f144dcc6e3b892885955df5572a).